qos-monitor/src/main/java/eu/arrowhead/core/qos/service/QoSMonitorDriver.java

```diff
@@ -164,9 +196,160 @@ public void requestGatekeeperInitRelayTest(final QoSRelayTestProposalRequestDTO
+
+		final UriComponents orchestrationProcessUri = getOrchestrationProcessUri();
+		try {
+
+			final ResponseEntity<OrchestrationResponseDTO> response = httpService.sendRequest(orchestrationProcessUri, HttpMethod.POST, OrchestrationResponseDTO.class, form);
+
+			return response.getBody();
+
+		} catch (final Exception ex) {
+			logger.debug("QoS Monitor can't access Orchestrator : " + ex.getMessage());
+
+			throw new ArrowheadException("QoS Monitor can't access Orchestrator.");
+		}
+
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public void checkPingMonitorProviderEchoUri(final UriComponents echoUri) {
+		logger.debug("checkPingMonitorProviderEchoUri started...");
+
+		int count = 0;
+
+		while ( count < MAX_RETRIES) {
+
+			rest();
+
+			try {
+
+				httpService.sendRequest(echoUri, HttpMethod.GET, String.class);
+
+				return;
+
+			} catch (final Exception ex) {
+				logger.warn("QoS Monitor can't access External Qos Monitor provider at : " + echoUri );
+
+				count++;
+
+			}
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public void subscribeToExternalPingMonitorEvents(final SystemRequestDTO pingMonitorProvider) {
+		logger.debug("subscribeToExternalPingMonitorEvents started...");
+
+		int count = 0;
+		boolean subscribedToAll = false; 
+
+		while ( !subscribedToAll && count < MAX_RETRIES) {
+
+			try {
+
+				final SubscriptionRequestDTO subscriptionRequest = getSubscriptionRequestDTOForExternalPingMonitor(pingMonitorProvider);
+				final UriComponents subscriptionUri = getEventHandlerSubscribeUri();
+
+				for (final QosMonitorEventType externalPingMonitorEventType : QosMonitorEventType.values()) {
+					subscriptionRequest.setEventType(externalPingMonitorEventType.name());
+
+					final ResponseEntity<SubscriptionResponseDTO> response = httpService.sendRequest(subscriptionUri, HttpMethod.POST, SubscriptionResponseDTO.class, subscriptionRequest);
+
+				}
+
+				subscribedToAll = true;
+
+			} catch (final Exception ex) {
+				logger.debug("QoS Monitor can't access EventHandler : " + ex.getMessage());
+
+				count++;
+
+				if (count < MAX_RETRIES) {
+					logger.debug("Retrying to access EventHandler.");
+
+					rest();
+				}
+			}
+		}
+
+		if (!subscribedToAll) {
+
+			throw new ArrowheadException("QoS Monitor can't subscribe to required events.");
+		}
+
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public void unsubscribeFromPingMonitorEvents() {
+		logger.debug("unSubscribeFromPingMonitorEvents started...");
+
+
+
+		int count = 0;
+		boolean unsubscribedFromAll = false; 
+
+		while ( !unsubscribedFromAll && count < MAX_RETRIES) {
+			try {
+				for (final QosMonitorEventType externalPingMonitorEventType : QosMonitorEventType.values()) {
+
+					final UriComponents unsubscriptionUri = getEventHandlerUnsubscribeUri().expand(Map.of(
```

Comment:
you can create the map before the loop, add all static key-value pairs once, and here just switch the event type

Reply:
fixed